packages/react-mongo/react-mongo.ts

```diff
@@ -0,0 +1,206 @@
+  params: {
+    name?: string,
+    args: any[]
+  }
+}
+
+const useSubscriptionClient = (name?: string, ...args: any[]): Meteor.SubscriptionHandle => {
+  const forceUpdate = useForceUpdate()
+
+  const refs: useSubscriptionRefs = useRef({
+    updateOnReady: false,
+    isReady: false,
+    params: {
+      name,
+      args
+    }
+  }).current
+
+  if (!EJSON.equals(refs.params, { name, args })) {
+    refs.updateOnReady = false
+    refs.isReady = false
+    refs.params = { name, args }
+  }
+
+  useEffect(() => {
+    const computation = Tracker.nonreactive(() => (
+      Tracker.autorun(() => {
+        const { name, args } = refs.params
+        if (!name) return
+
+        refs.subscription = Meteor.subscribe(name, ...args)
+
+        const isReady = refs.subscription.ready()
+        if (isReady !== refs.isReady) {
+          refs.isReady = isReady
+          if (refs.updateOnReady) {
+            forceUpdate()
+          }
+        }
+      })
+    ))
+
+    return () => {
+      computation.stop()
+      delete refs.subscription
+    }
+  }, [refs.params])
+
+  return {
+    stop () {
+      refs.subscription?.stop()
+    },
+    ready () {
+      // Ready runs synchronously with render, should not create side effects.
+      refs.updateOnReady = true
+      return refs.isReady
+    }
+  }
+}
+
+const useSubscriptionServer = (name?: string, ...args: any[]): Meteor.SubscriptionHandle => ({
+  stop() {},
+  ready() { return true }
+})
+
+export const useSubscription = Meteor.isServer
+  ? useSubscriptionServer
+  : useSubscriptionClient
+
+type useFindActions<T> =
+  | { type: 'refresh', data: T[] }
+  | { type: 'addedAt', document: T, atIndex: number }
+  | { type: 'changedAt', document: T, atIndex: number }
+  | { type: 'removedAt', atIndex: number }
+  | { type: 'movedTo', fromIndex: number, toIndex: number }
+
+const useFindReducer = <T>(data: T[], action: useFindActions<T>): T[] => {
+  switch (action.type) {
+    case 'refresh':
+      return action.data
+    case 'addedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        action.document,
+        ...data.slice(action.atIndex)
+      ]
+    case 'changedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        action.document,
+        ...data.slice(action.atIndex + 1)
+      ]
+    case 'removedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        ...data.slice(action.atIndex + 1)
+      ]
+    case 'movedTo':
+      const doc = data[action.fromIndex]
+      const copy = [
+        ...data.slice(0, action.fromIndex),
+        ...data.slice(action.fromIndex + 1)
+      ]
+      copy.splice(action.toIndex, 0, doc)
+      return copy
+  }
+}
+
+const useFindClient = <T = any>(factory: () => Mongo.Cursor<T>, deps: DependencyList) => {
+  const [data, dispatch] = useReducer<Reducer<T[], useFindActions<T>>, T[]>(
+    useFindReducer,
+    undefined,
+    () => Tracker.nonreactive(() => {
+      const cursor = factory()
```

Comment:
We should validate for a cursor here, (we can elaborate on the messaging a little).

```suggestion
      const cursor = factory()
      if (!(cursor instanceof Mongo.Cursor)) {
        throw Error('Not a cursor.')
      }
```